nmpolicy/internal/resolver/replace.go

```diff
@@ -41,14 +34,55 @@ func replace(inputState map[string]interface{}, path ast.VariadicOperator, repla
 	return replacedMap, nil
 }
 
-func replaceMapFieldValue(replaceValue interface{}) mapEntryVisitFn {
-	return func(inputMap map[string]interface{}, mapEntryKeyToReplace string) (interface{}, error) {
-		modifiedMap := map[string]interface{}{}
-		for k, v := range inputMap {
-			modifiedMap[k] = v
-		}
+type replaceOpVisitor struct {
+	replaceValue interface{}
+}
+
+func (r *replaceOpVisitor) visitLastMap(p path, inputMap map[string]interface{}) (interface{}, error) {
+	modifiedMap := map[string]interface{}{}
+	for k, v := range inputMap {
+		modifiedMap[k] = v
+	}
 
-		modifiedMap[mapEntryKeyToReplace] = replaceValue
-		return modifiedMap, nil
+	modifiedMap[*p.currentStep.Identity] = r.replaceValue
+	return modifiedMap, nil
+}
+
+func (*replaceOpVisitor) visitLastSlice(path, []interface{}) (interface{}, error) {
+	return nil, nil
+}
+
+func (r *replaceOpVisitor) visitMap(p path, mapToVisit map[string]interface{}) (interface{}, error) {
+	interfaceToVisit, ok := mapToVisit[*p.currentStep.Identity]
+	if !ok {
+		return nil, nil
+	}
+
+	visitResult, err := visitState(p.nextStep(), interfaceToVisit, r)
+	if err != nil {
+		return nil, err
+	}
+
+	replacedMap := map[string]interface{}{}
+	for k, v := range mapToVisit {
+		replacedMap[k] = v
+	}
+	replacedMap[*p.currentStep.Identity] = visitResult
+	return replacedMap, nil
+}
+
+func (r *replaceOpVisitor) visitSlice(p path, sliceToVisit []interface{}) (interface{}, error) {
+	if p.currentStep.Number != nil {
```

Comment:
I think both visitSlice and visitMap should return an error if identity is nil.

Reply:
ditto